### PR TITLE
chore(internal/ssm): added 15 fuzzing tests.

### PR DESCRIPTION
**Pull request by MozamilS**
## Summary
Went through the files in `internal/ssm` with AI and generated 15 fuzzing checks. Almost all fuzzing tests went smoothly except a few and the issues were addressed in the relevant files. 

## Related issue
Work for #120 

## What changed
Added following fuzzers: 

1. **FuzzShellQuote** — Tests shell argument quoting for Unix/Linux systems. Validates proper escaping of special characters and edge cases like empty strings, spaces, and single quotes.

2. **FuzzPowerShellQuote** — Tests PowerShell argument quoting for Windows systems. Ensures correct escaping of PowerShell metacharacters and proper handling of quotes and variable expansion.

3. **FuzzSanitizeBasename** — Tests filename sanitization logic. Validates removal of directory separators and path traversal sequences while preserving valid filename characters.

4. **FuzzRemoteBaseName** — Tests remote path basename extraction. Ensures correct extraction of filenames from both Unix and Windows remote paths with various path formats.

5. **FuzzParseArg** — Tests command-line argument parsing for remote paths. Validates correct identification of remote hosts, Windows drive letters, and local paths with edge cases.

6. **FuzzNormalizeWindowsPath** — Tests Windows path normalization. Ensures consistent forward-slash to backslash conversion while handling edge cases like UNC paths and mixed separators.

7. **FuzzAllInstanceIDs** — Tests EC2 instance ID extraction from AWS API responses. Validates correct extraction from nested Reservations and Instances structures with nil safety.

8. **FuzzNameTag** — Tests EC2 instance name tag extraction. Ensures proper retrieval of the "Name" tag from instance tags with fallback handling for missing or malformed tags.

9. **FuzzTargetInfoFromInstance** — Tests target information assembly from EC2 instances. Validates instance ID precedence (instance ID over fallback) and platform preservation.

10. **FuzzFirstInstance** — Tests first instance retrieval from AWS API responses. Validates correct behavior with nil outputs, empty reservations, and proper instance selection.

11. **FuzzNoInstancesFound** — Tests empty instance detection. Ensures correct identification of nil or empty instance lists from AWS API responses.

12. **FuzzListInstancesFiltering** — Tests instance filtering logic. Validates case-insensitive name substring matching and platform equality matching with proper filter composition.

13. **FuzzDownloadBase64Decoding** — Tests base64 decoding for remote file downloads. Ensures proper handling of whitespace, invalid base64, and round-trip encoding/decoding integrity.

14. **FuzzParseRemoteFlag** — Tests remote flag parsing logic. Validates correct identification of remote instance specifications, handling of various flag formats, and proper fallback behavior for invalid or malformed inputs.

15. **FuzzParseS3URL** — Tests S3 URL parsing for remote file transfers. Ensures correct extraction of bucket names and object keys from S3 URLs with various formats, and validates round-trip integrity when reconstructing URLs from parsed components.

## Testing performed
make test
make lint

## Checklist
- ✅  I linked the related issue when applicable
- ✅  I reviewed testing standards in CONTRIBUTING.md
- ✅  I ran lint and formatting checks or explained why skipped
- ✅  I updated docs/comments if needed
- ✅  I followed commit conventions
- ✅  This PR is focused on one logical change

### CODE REVIEW

**Comment by MozamilS** [comment]
Trying to register the fuzzers to OSS Fuzz in this commit: https://github.com/google/oss-fuzz/pull/15744

I did register for CLA but I think I'll have to update the commit in a few hours so the system updates.

**Comment by rhysmcneill** [comment]
Hey @MozamilS 

How is registration for fuzzers going? Is this PR dependent on the fuzzer merge in Google repo or can we go ahead without it?

**Comment by MozamilS** [comment]
@rhysmcneill 

Hey there,

We should go ahead and merge this commit to the codebase.

I have opened a PR at oss-fuzz but it didn't go through because it was my first commit there and they need to manually check my CLA agreement. 

Once the PR gets merged there, they'll look at the codebase to find these fuzzers so they have to get added (unless you find any issues). 

I tried to trigger their automated CLA-check workflow yesterday but it still didn't go through. Can't do anything except wait.

**Comment by rhysmcneill** [comment]
Great @MozamilS 

I will review this ticket now :)

File: internal/ssm/download_base64_fuzz_test.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+				t.Errorf("FuzzDownloadBase64Decoding: re-encoded string failed to decode: %v", err)
+			}
+			if string(decoded) != string(redecoded) {
+				t.Errorf("FuzzDownloadBase64Decoding: round-trip mismatch for %q", encoded)
+			}
+			return
+		}
+
+		// Invariant: TrimSpace removes leading/trailing whitespace correctly
+		if strings.TrimSpace(encoded) != trimmed {
```
Comment (rhysmcneill) [inline internal/ssm/download_base64_fuzz_test.go]: **Must fix — this invariant can never fail**

`trimmed` was assigned from `strings.TrimSpace(encoded)` just above, so `strings.TrimSpace(encoded) != trimmed` is always `false`. The fuzzer will never trip this regardless of what it throws at it, so it's giving false confidence.

Since this branch only runs when decoding failed, something more useful would be to assert the error is actually non-nil:

```go
// decoding failed — ensure we got an error back
if err == nil {
    t.Errorf("FuzzDownloadBase64Decoding: expected decode error for %q but got nil", encoded)
}
```

Or just remove the block entirely if there's nothing meaningful to assert here.

File: internal/ssm/remote_fuzz_test.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	f.Fuzz(func(t *testing.T, input string) {
+		host, port, err := ParseRemoteFlag(input)
+
+		// Invariant: if error is nil, port must be valid (1-65535)
+		if err == nil {
+			if port < 1 || port > 65535 {
+				t.Errorf("ParseRemoteFlag(%q) returned invalid port %d", input, port)
+			}
+			// If no error and RemoteHost is set, it should not be empty
+			if host == "" || host != "" {
```
Comment (rhysmcneill) [inline internal/ssm/remote_fuzz_test.go]: **Must fix — dead code, condition is always true**

`host == "" || host != ""` is just `true`, so this block always runs and the `_ = host` no-op is all it does. The comment even admits it. Either replace it with a real invariant or remove it.

For example, when there's no error you could check that if a host was parsed it doesn't still contain a colon (since the port should have been split out):

```go
if host != "" && strings.Contains(host, ":") {
    t.Errorf("ParseRemoteFlag(%q): host %q still contains a colon after parsing", input, host)
}
```

File: internal/ssm/list_instances_filtering_fuzz_test.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	f.Add("prod", "linux")        // both filters
+	f.Add("DB", "WINDOWS")        // case variations
+	f.Add("nonexistent", "linux") // name filter matches nothing
+	f.Add("app", "unsupported")   // platform filter matches nothing
+
+	f.Fuzz(func(t *testing.T, nameFilter, platformFilter string) {
+		items := generateSSMInstances()
+		names := generateNameMap()
+
+		var result []InstanceInfo
```
Comment (rhysmcneill) [inline internal/ssm/list_instances_filtering_fuzz_test.go]: **Nice to have — the fuzz body reimplements the filtering rather than calling the real function**

The loop here duplicates the filtering logic inline instead of calling `listInstances`. The invariants you're checking are sensible, but they're being verified against a copy of the logic rather than the actual production code. If someone introduces a bug into `listInstances` down the line, this test won't catch it.

If calling `listInstances` directly is awkward in a fuzz context (e.g. it needs an AWS client), one option is to extract the filtering predicate into a small pure function that can be tested in isolation.

File: internal/ssm/all_instance_ids_fuzz_test.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import (
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+)
+
+func FuzzAllInstanceIDs(f *testing.F) {
+	f.Add(0)  // nil input
```
Comment (rhysmcneill) [inline internal/ssm/all_instance_ids_fuzz_test.go]: **Nice to have — integer scenario selector limits what the fuzzer can explore**

Using `int` as the fuzz input and switching on it means the fuzzer is only mutating an integer across a small fixed set of cases. It can't discover unexpected struct shapes — things like a reservation with a nil `InstanceId`, a mix of nil and non-nil instances, or a very large number of reservations.

Same pattern applies to `FuzzNameTag`, `FuzzFirstInstance`, and `FuzzTargetInfoFromInstance`.

Fuzzing the primitive fields that actually drive behaviour gives the engine a much larger search space, e.g.:

```go
f.Fuzz(func(t *testing.T, numReservations int, instancesPerRes int, includeNilID bool) {
    // build output dynamically
})
```

File: internal/ssm/remote_basename_fuzz_test.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package ssm
+
+import (
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
```
Comment (rhysmcneill) [inline internal/ssm/remote_basename_fuzz_test.go]: **Nice to have — import ordering**

Stdlib imports should come before third-party ones (`goimports` enforces this):

```go
import (
    "testing"

    "github.com/aws/aws-sdk-go-v2/service/ec2/types"
)
```

File: internal/ssm/normalize_windows_path_fuzz_test.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+				inputBackslashes++
+			}
+		}
+		resultBackslashes := 0
+		for i := 0; i < len(result); i++ {
+			if result[i] == '\\' {
+				resultBackslashes++
+			}
+		}
+		if resultBackslashes < inputBackslashes {
```
Comment (rhysmcneill) [inline internal/ssm/normalize_windows_path_fuzz_test.go]: **Nice to have — invariant only checks backslashes weren't lost, not that forward slashes were converted**

The current check passes as long as `resultBackslashes >= inputBackslashes`, but it doesn't verify that the forward slashes from the input actually became backslashes in the output. A tighter version:

```go
inputForwardSlashes := strings.Count(input, "/")
// each / should become \, existing \ should be preserved
expectedBackslashes := inputBackslashes + inputForwardSlashes
if resultBackslashes != expectedBackslashes {
    t.Errorf("normalizeWindowsPath(%q): expected %d backslashes, got %d", input, expectedBackslashes, resultBackslashes)
}
```

**Review by rhysmcneill — CHANGES_REQUESTED** [review]
Nice work on this — the two real bugs found through fuzzing (`ParseArg` and `sanitizeBasename`) are genuinely useful catches and the fixes look correct. Left a couple of must-fix comments on invariants that can never actually fire, and a few nice-to-haves where the tests could be tightened up.

File: internal/ssm/remote_fuzz_test.go (hunk 2 above)
Comment (MozamilS) [reply internal/ssm/remote_fuzz_test.go]: Found an issue after the condition above was added to the fuzzer. 

Cases such as `::1`, `a:b:5432` were getting passed. 

Remedied this with adding a condition in `forward.go` to check whether `host` has any `:` (as per the condition in the fuzzer). 

Investigation lead to the `TestParseRemoteFlag_MultipleColons` test in `forward_test.go` where `a:b:5432` was actually getting passed. 

Remedied this with renaming the test to `TestParseRemoteFlag_DottedHostname` to validate a case like `db.example.com:5432`. 

All test pass.

File: internal/ssm/list_instances_filtering_fuzz_test.go (hunk 3 above)
Comment (MozamilS) [reply internal/ssm/list_instances_filtering_fuzz_test.go]: Applied the small pure function approach and made `filterInstances` in`list.go`. A small issue was caught with empty ids that was fixed. 

Later, `all_instance_ids_fuzz_test.go` also started throwing an error that seemed related to the empty id issue that was also fixed.

**Comment by MozamilS** [comment]
@rhysmcneill 

Hello there, 

I brought in the changes requested above. Hopefully everything is covered sufficiently.

File: internal/ssm/download_base64_fuzz_test.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+				t.Errorf("FuzzDownloadBase64Decoding: re-encoded string failed to decode: %v", err)
+			}
+			if string(decoded) != string(redecoded) {
+				t.Errorf("FuzzDownloadBase64Decoding: round-trip mismatch for %q", encoded)
+			}
+			return
+		}
+
+		// decoding failed — ensure we got an error back
+		if err == nil {
```
Comment (rhysmcneill) [inline internal/ssm/download_base64_fuzz_test.go]: **Must fix — this check is still dead code, just in a different form**

The `if err == nil` block on line 28 always calls `return` before reaching this point, so by the time we get here `err` is guaranteed to be non-nil. The `if err == nil` guard on this line can therefore never be true — it's the same structural problem as before.

The simplest fix is to just remove the guard entirely, since we're already in the error path:

```go
// decoding failed — nothing further to assert here
```

Or, if you want an affirmative assertion, flip the condition:

```go
// Invariant: we must have reached here via the error path
if err == nil {
    t.Errorf(...) // this will never fire — just delete the block
}
```

The cleanest option is to simply drop the block. The round-trip invariant in the success branch is the meaningful check here.

**Review by rhysmcneill — CHANGES_REQUESTED** [review]
Good progress — the two real bugs in `ParseArg` and `sanitizeBasename` are fixed, the `filterInstances` refactor is clean and caught another real empty-ID bug, and the `ParseRemoteFlag` invariant improvement found and fixed the IPv6/multi-colon case. Most things are in good shape.

One must-fix remains in `download_base64_fuzz_test.go` — the error check is still dead. Also flagging `FuzzFirstInstance` and `FuzzNameTag` as nice-to-haves since they still use the integer scenario pattern that was already improved in `FuzzAllInstanceIDs` and `FuzzTargetInfoFromInstance`.

File: internal/ssm/download_base64_fuzz_test.go (hunk 7 above)
Comment (MozamilS) [reply internal/ssm/download_base64_fuzz_test.go]: @rhysmcneill 

Hello, 

This code block was dropped. I double checked the fuzzer to see if it's still passing just to be sure and everything's fine as expected.

**Review by rhysmcneill — APPROVED** [review]
All must-fixes addressed — the dead error check in `download_base64_fuzz_test.go` has been removed. The remaining `FuzzFirstInstance` and `FuzzNameTag` integer-selector pattern is a nice-to-have and not a blocker. Good to merge.
